reframe/core/schedulers/__init__.py

```diff
@@ -14,12 +13,49 @@
 from reframe.core.logging import getlogger
 
 
-class Job(abc.ABC):
+class JobScheduler(abc.ABC):
+    @abc.abstractmethod
+    def emit_preamble(self, job):
+        pass
+
+    @abc.abstractmethod
+    def allnodes(self):
+        '''Gets all the available nodes'''
+
+    @abc.abstractmethod
+    def filternodes(self, job, nodes):
+        '''Filter nodes according to the job options'''
+
+    @abc.abstractmethod
+    def submit(self, job):
+        pass
+
+    @abc.abstractmethod
+    def wait(self, job):
+        pass
+
+    @abc.abstractmethod
+    def cancel(self, job):
+        pass
+
+    @abc.abstractmethod
+    def finished(self, job):
+        pass
+
+
+class Job:
```

Comment:
I really liked the design. The separation of `Job`  into `JobScheduler` and `Job` allows a simpler implementation of the different schedulers and their functionalities. Now, `Job` is a struct that holds info about the job state, which is populated by the scheduler that schedules it.